Feat/slider Slider 컴포넌트 구현

iyu88 opened the pull request:
## 🤠 개요

- #107 
- 단일 Thumb를 가지고 있는 Slider 컴포넌트를 구현합니다. 

## 💫 설명

- 기존에 있던 RangeSelector(단일 Thumb 버전)에서 이름만 바뀌고 역할은 동일한 컴포넌트인 Slider 를 구현했습니다. 
  - 추후 이중 Thumb를 가지고 있는 Slider는 RangeSlider 라는 이름으로 구현할 예정입니다. 

- 왜 단일부터 만들었냐면 RangeSelector의 구현 방식이 처음부터 안좋은 방식으로 되어 있던 것 같아서 그랬습니다. 
  - 대표적으로 useLayoutEffect에서 DOM style 속성을 설정하여 초기 Filled (Slider에서 채워진 부분)와 
    Thumb의 위치를 결정하는 로직입니다.
  - 또한 Thumb를 조절할 때 DOM의 위치와 value를 계산하는 로직을 일일히 짜다보니까 수식에 실수가 들어가기도 하고 
    렌더링 오류로 직결되는 점이 걸렸습니다. 

- 새로운 컴포넌트라고 봐주시면 좋겠지만, 기존의 RangeSelector와 다른 점을 정리하면 다음과 같습니다. 

---

- Track 부분과 Thumb가 강결합되어 있던 점을 하위 컴포넌트 3가지 Track, Filled, Thumb 으로 분리했습니다. 
   - Track : Thumb가 움직일 수 있는 범위입니다. 
   - Filled : 현재 value만큼 Track이 채워지는 것을 표현하는 컴포넌트입니다. 
   - Thumb : Slider의 컨트롤을 담당합니다. 
 
- useSlider 라는 커스텀 훅에 위치 계산 로직이나 스타일 관련 로직을 위치시켰습니다. 
  - 앞에서 'Track 부분과 Thumb가 강결합되어 있다' 고 했는데 컴포넌트는 분리하고 계산 로직만 묶었다고 보시면 될 것 같습니다. 
  - 결국 Track, Filled, Thumb는 긴밀하게 동작할 수 밖에 없기 때문에 계산 로직을 커스텀 훅에 담았고, 
     동일한 이유로 함께 변경되는 CSS 스타일 객체도 훅 내부에 위치시켰습니다. 

- 앞서 설명드린 것처럼 위치 계산 방법을 변경했습니다. 
  - position: absolute인 점을 이용하여 비율 (%) 를 계산해서 left 나 bottom 을 설정하기 때문에 위치 오류가 발생하지 않습니다.
     (특히, resize 시)

- 방향 설정이 가능합니다 ( default 는 horizontal )
  
- step 으로 한 번에 조절되는 value를 설정할 수 있습니다.  ( default 는 1 )

- 키보드 접근성에 디테일을 추가했습니다. [참고 링크 - 널리](https://nuli.navercorp.com/community/article/1132993)
  - ⬅️ ⬇️ : 값이  step만큼 감소합니다. 
  - ➡️ ⬆️ : 값이 step만큼 증가합니다.
  - PageDown :  10% 감소합니다.
  - PageUp : 10% 증가합니다.  
  - Home : 최소값이 됩니다. 
  - End : 최대값이 됩니다. 
    
- FormData에서 값을 받아올 때를 대비해서 input 태그를 Thumb 내부에 위치시켰습니다. 

- 자세한 동작은 스토리북을 확인해주세요

## 📷 스크린샷 (Optional)

- 키보드 이벤트 영상 첨부합니다. 

### 방향키

https://user-images.githubusercontent.com/31645195/233601353-ceee384e-a2a6-4c1e-af33-4fbf6714e1af.mov

### PageDown / PageUp

https://user-images.githubusercontent.com/31645195/233601421-abe3fe55-0d88-4f6c-a62f-c5d3c9aecd1f.mov

### Home / End

https://user-images.githubusercontent.com/31645195/233601409-71c503ec-9849-422d-9afe-f58d773a848b.mov

iyu88 committed:
feat: RangeSelector props 수정

id props은 삭제하고 label prop을 활용합니다.
방향을 결정하는 orientation prop을 추가합니다.

iyu88 committed:
feat: useSafeContext 적용

useContext를 추상화한 useSafeContext를 적용합니다.

iyu88 committed:
feat: Track 과 Filled 를 하위 컴포넌트로 분리

iyu88 committed:
feat: Thumb 를 하위 컴포넌트로 분리

iyu88 committed:
feat: Position 속성으로 전체 Slider 위치 조정

position을 absolute로 설정하여 각 하위 컴포넌트의 위치를 적절하게 설정합니다.

iyu88 committed:
feat: Thumb 마우스 이동 이벤트 설정

iyu88 committed:
refactor: 하위 컴포넌트 스타일 객체 변수명 변경

iyu88 committed:
feat: useSlider 커스텀 훅에 마우스 이벤트 로직 분리

마우스 이벤트를 실행했을 때 Thumb와 Filled를 적절하게 움직이기 위한 이벤트 핸들러를 분리합니다.

iyu88 committed:
feat: Slider에서 사용할 CSS 객체 선언

스타일 객체를 각 하위 컴포넌트에서 import 해서 사용하기 위해서 커스텀 훅 내부에 스타일을 가져올 수 있는 함수를 생성합니다.
Thumb의 위치와 Filled의 비율을 해당 함수에 바로 적용하기 위해서 각각을 저장하기 위한 state를 추가로 선언합니다.

iyu88 committed:
feat: 거리 비율 계산식 수정 및 초기값 설정

iyu88 committed:
feat: Thumb 이동 범위 제한

iyu88 committed:
feat: step prop에 대응하며 값 변경

iyu88 committed:
feat: onKeyDown 이벤트 핸들러 구현

iyu88 committed:
feat: 마우스 이벤트 핸들러 이름 변경

기존에 있던 마우스 이벤트 핸들러 관련 함수들을 통폐합합니다.
최종적인 함수는 onMoveSlider 만 남겨서 마우스 이벤트를 핸들링합니다.

iyu88 committed:
refactor: useSlider 코드 스타일 정리

iyu88 committed:
feat: Slider로 컴포넌트 생성

Slider로 컴포넌트 이름을 변경하고 useSlider 훅을 적용합니다.

iyu88 committed:
feat: Slider 스토리 정리

iyu88 committed:
fix: Slider 스토리 argTables 수정

iyu88 committed:
feat: Slider Root의 불필요한 css 속성 삭제

iyu88 committed:
feat: Slider 접근성 속성 추가

기존에 잘못되었던 접근성 속성 사용법을 바로잡습니다.

https://nuli.navercorp.com/community/article/1132993

iyu88 committed:
feat: KeyDown 이벤트 핸들러에 키 추가

orientation에 따라 화살표 키로 값을 조절하는 방식이 달라졌던 점을 orientation과 무관하게 수정되도록 합니다.
추가적인 PageUp / PageDowm / Home / End 키에 이벤트를 추가합니다.

- PageDown : 10% 줄입니다.
- PageUp : 10% 높입니다.
- Home : 최소값으로 만듭니다.
- End : 최대값으로 만듭니다.

https://nuli.navercorp.com/community/article/1132993
https://www.freecodecamp.org/news/javascript-keycode-list-keypress-event-key-codes/

iyu88 committed:
style: Slider 컴포넌트 코드 스타일 수정

iyu88 committed:
feat: hidden input 태그 추가

FormData를 사용해서 Slider 값을 가져오게 될 경우에 대비하여 input 태그를 추가합니다.
display: none 으로 설정해서 렌더링에 방해되지 않도록 합니다.

iyu88 committed:
feat: 배포용 경로 변경

prayinforrain reviewed (approved):
멋있어요~ 처음 것보다 코드 읽기도 훨씬 수월해졌어요. 슬라이더 자체가 어려운 컴포넌트지만 ㅎㅎ..
몇가지 질문과 제안이 있어요.

1. Track의 clickable 영역이 조금 더 넓었으면 해요.
트랙은 보통 가느다란(?) 편이니까.. 적어도 Thumb가 차지하는 만큼의 높이(수직 슬라이더면 너비)는 가져가도 되지 않을까요? padding과 pointer-events를 잘 넣으면 될 것 같아요(추상적)
2. 사용에 꼭 합성 컴포넌트가 필요할까요?
제가 이해한 구조가 맞다면 color는 Slider 컴포넌트의 props로 받아서 내려줘도 될 것 같고, custom thumb도 Slider props에 추가해도 괜찮을 것 같아요. 뭔가 더 복잡한 커스터마이징 시나리오가 있다면 지금의 형태로 분리해서 사용하는 게 이득이겠지만 지금은 분리해서 구현되어 있더라도 사용자에게는 하나의 `<Slider/>` 형태로 합쳐서 제공해도 될 것 같아요. 다른 이유가 있을 수도 있겠지만요 ㅎㅎ..
3. size를 `100%`로 주는 방법은 없을까요?
기술적으로 어려움이 있을 수도 있겠지만 제가 만약 사용자라면 이게 가능한지를 찾아볼 것 같아요.
4. `step` props의 단위는 px인가요 value인가요?
withStep10 스토리를 보면 width가 200px이고, step이 10이고, value값은 0 ~ 100으로 보이는데, 설명에 의하면 step : 10은 10px씩 움직이니 value는 5단위로 조절되어야 할 것처럼 생각되지만 실제로는 10씩 움직여요.

그리고.. 드래그를 지나치게 빠르게 하면 value가 끝값까지 안오는 버그가 있네요. 이건 해결할 수 있는 문제인지는 잘 모르겠어요. 아니면 제 컴퓨터가 지금 아픈걸수도(?)

leesunmin1231 commented on src/components/Slider/index.tsx:
```diff
@@ -0,0 +1,232 @@
+  min,
+  max,
+  defaultValue,
+  size,
```
이 size는 Track의 가로 길이인 거 같은데 맞나요? default value가 지정되어 있으면 더 쓰기 편할거 같아요~

iyu88 replied:
이거는 우재님이 말씀하신 것처럼 100%로 사용할 수 있는 방향이나 
선민님이 말씀하신 것처럼 기본 값을 주는 방식으로 수정해볼게요.
vertical인 경우에는 어떻게 할지도 생각해볼게요.
의견 주시면 감사하겠습니다~

iyu88 replied:
size의 default 를 100%로 받도록 설정했습니다.

leesunmin1231 commented on src/components/Slider/useSlider.ts:
```diff
@@ -0,0 +1,170 @@
+    e.preventDefault();
+
+    let nextValue = (function (key: string, currentValue: number) {
+      if (key === ARROW_LEFT || key === ARROW_DOWN) {
+        return currentValue - step;
+      } else if (key === ARROW_RIGHT || key === ARROW_UP) {
+        return currentValue + step;
+      } else if (key === PAGE_DOWN) {
+        // decrease 10%
+        return currentValue - (max - min) / 10;
+      } else if (key === PAGE_UP) {
+        // increase 10%
+        return currentValue + (max - min) / 10;
+      } else if (key === HOME) {
+        return min;
+      } else if (key === END) {
+        return max;
+      }
```
이건 그냥 제 취향이긴 한데, switch를 쓰는건 어때요? 저는 그게 더 깔끔해 보이더라구요 ㅎㅎ 무시하셔도 돼요!

iyu88 replied:
네 반영해볼게요~

leesunmin1231 commented on src/components/Slider/useSlider.ts:
```diff
@@ -0,0 +1,170 @@
+  const [value, setValue] = useState<number>(defaultValue);
+
+  const defaultRatio = Math.round(((defaultValue - min) / (max - min)) * 100);
+  const [thumbPosition, setThumbPosition] = useState<number>(defaultRatio);
+  const [filledRatio, setFilledRatio] = useState<number>(defaultRatio);
```
이 두 상태가 계속 같은 값을 사용하는 거 같은데 맞나요? 맞다면 그냥 `[ratio, setRatio]`하나의 상태로 관리하는건 어때요?

iyu88 replied:
그러게요 ㅎㅎ 원래는 둘의 목적이 달랐는데
어느 순간 하나가 되었네요
변경해보고 의도대로 동작하는지 확인해볼게요

leesunmin1231 commented on src/components/Slider/useSlider.ts:
```diff
@@ -0,0 +1,170 @@
+  };
+
+  const onMoveSlider = (
+    e: MouseEvent<HTMLDivElement> | globalThis.MouseEvent,
```
헉 globalThis 이건 뭐죠? 처음봐요

se030 replied:
브라우저에서는 window가 될텐데 없어도 되는 네임스페이스라고 생각해요 ~

iyu88 replied:
이거는 내일 회의 때 공유할게요

se030 replied:
지금보니 ... 타입명이 중복이네요 ... 근데 저는 globalThis보다 window로 쓰는게 더 좋을 것 같아요. 근데 안주무시나요?

iyu88 replied:
```suggestion
    e: ReactMouseEvent<HTMLDivElement> | MouseEvent,
```

이렇게 변경하겠습니다 감사합니다~

<br />

**_jsx에 붙일 때는 react 이벤트로 타이핑이 되어 있어야 하고_** 
_**DOM에 직접 붙일 때는 native 이벤트여야 한다**_
_**- se030, prayinforrain -**_

leesunmin1231 reviewed (approved):
고생하셨습니다~!! 위치 계산로직을 훅으로 따로 빼니까 훨씬 깔끔해요 ㅎㅎ
Thumb 안에 숫자는 확인용으로 넣어두신 건가요? default 스타일로 심플한 동그라미 하나 달아 놓으면 예쁠거 같아요


수정) 제것도 우재님이 말하신 것과 같은 버그가 있어요. value가 끝값까지 안오네요..

se030 commented on src/components/Slider/useSlider.ts:
```diff
@@ -0,0 +1,170 @@
+    const { left, right, bottom, top } =
+      trackRef.current.getBoundingClientRect();
+
+    // Verify that the event point is valid
+    if (isHorizontal) {
+      if (e.clientX < left || e.clientX > right) return;
+    } else {
+      if (e.clientY > bottom || e.clientY < top) return;
+    }
```
다른 분들이 말씀해주신 버그가 이 부분 때문인 것 같은데, Math.min, max로 범위 내로 보정해주면 어떨까요?

iyu88 replied:
그렇네요 이 부분 수정하면 될 것 같아요 👍

se030 commented on src/components/Slider/Slider.stories.tsx:
```diff
@@ -0,0 +1,258 @@
+} as ComponentMeta<typeof Slider>;
+
+const DEFAULT_LABEL = 'cds_Slider';
+const DEFAULT_PROPS = {
```
전부 동일한 값을 사용한다면 Template 자체에 props 달아주면 어떤가요?

iyu88 replied:
DEFAULT_PROPS 객체 내부에서 사용하고 있는 값이니까 괜찮을 것 같다고 생각했어요 

혹, label만 상수로 분리되어 있는게 조금 이상해보이나요?

label: 'cds_Slider'

로 수정하는 방향도 있을 것 같네요.

iyu88 replied:
고치려고 보니까 Template이 여러 개이기도 하고 

위에서 말씀드린 것처럼 DEFAULT_PROPS 객체를 이미 공통적으로 전달하고 있어서 

지금처럼 해당 객체 내부에서 사용하는 방향으로 유지하겠습니다!

se030 commented on src/components/Slider/index.tsx:
```diff
@@ -0,0 +1,232 @@
+
+interface TrackProps {
+  color?: string;
+  children: EmotionJSX.Element;
```
이건 왜 ReactNode가 아닌가요?

iyu88 replied:
그러네요 수정하겠습니다

se030 reviewed (approved):
고생하셨습니다! 계산 로직이 분리돼서 눈에 잘 들어왔어요 ㅎㅎ

저도 사용법 관련해서 하나 코멘트하면 사용처에서 Track, Filled, Thumb를 커스텀하고 싶지 않은 경우에는 하위 컴포넌트를 따로 작성하지 않아도 기본 스타일을 제공할 수 있으면 좋겠어요 ~ Slider에서 children 여부로 판단이 가능할 것 같아요.

iyu88 committed:
refactor: Type, props 변경

size의 기본값을 100% 로 받도록 변경합니다.
props 중에서 optional로 받아도 되는 값을 type에 명시합니다.

iyu88 committed:
refactor: Slider 스토리 수정

변경된 props와 잘못된 설명을 수정합니다.
Slider 태그만 사용한 템플릿을 추가합니다.

iyu88 committed:
refactor: 코드 리뷰 사항 반영

filled와 track에서 사용하는 상태를 하나로 합칩니다.
onPressArrow의 코드 스타일을 switch문으로 변경합니다.
onMoveSlider의 이벤트 타입에 native 이벤트도 추가합니다.
마우스 이벤트를 마우스 위치가 아닌 계산된 value 값을 기준으로 제한합니다.
변경된 size prop을 하위 컴포넌트들의 스타일 객체에 반영합니다.

iyu88 committed:
feat: Docs Show Code 수정

Show Code에 표시되는 내용을 깔끔하게 정리합니다.
하위 컴포넌트에 displayName을 부여합니다.
decorator에서 사용하는 컴포넌트의 스타일 정의 방식을 수정합니다.

iyu88 committed:
style: 즉시 실행 함수를 화살표 함수 형태로 변경

iyu88 committed:
feat: 스토리용 패딩 추가

Slider의 root에 있던 margin 값을 삭제합니다.
스토리 decorator에 padding을 추가합니다.

iyu88 committed:
Merge branch 'main' into feat/slider
